xapian-applications/omega/date.cc

```diff
@@ -144,7 +147,40 @@ date_range_filter(const string & date_start, const string & date_end,
 {
     int y1, m1, d1, y2, m2, d2;
     if (!date_span.empty()) {
-	time_t secs = atoi(date_span.c_str()) * (24 * 60 * 60);
+    	time_t secs;
+	// For x32 system use long (32bit)
+	// For x64 system use long long (64bit)
+	if (sizeof(int *) < 8) {
+	    long str_to_long;
+	    errno = 0;
+	    str_to_long = strtol(date_span.c_str(),NULL,0);
+	    if (errno == ERANGE) {
+		switch(str_to_long) {
+		    case LONG_MIN:
+			throw (string("Date span underflow: ") + string(strerror(errno)));
+		    break;
+		    case LONG_MAX:
+			throw (string("Date span overflow: ") + string(strerror(errno)));
+		    break;
+		}
+	    }
+	    secs = time_t(str_to_long) * 24 * 60 * 60;
```

Comment:
You carefully check for overflow in the conversion, but then blindly multiply here, which could easily overflow (e.g. if `DATESPAN=99999` and the time_t is 32 bit).

I think here it's not actually helpful to raise an error in either case.  It means that `DATESPAN=9999999999` will always raise an exception on some architectures, but work on others.  It would be more helpful to clamp overflow to the largest possible `time_t` value - we can't represent a time after that, so it really is the appropriate limit in this case.